### ISSUE TITLE
Improve resolution of canvas on HiDPI with PyQt5 backend

### ISSUE THREAD

**Issue by astrofrog** [matplotlib/matplotlib#4977]
One of the nice things about Qt5 (as far as I understand) is that it provides better support for HiDPI (retina) displays, but at the moment it doesn't look like Matplotlib is taking advantage of this:

**PyQt4**:

<img width="400" alt="pyqt4" src="https://cloud.githubusercontent.com/assets/314716/9424369/b9e13a74-48ea-11e5-830e-92297a819ec7.png">

**PyQt5:**

<img width="400" alt="pyqt5" src="https://cloud.githubusercontent.com/assets/314716/9424370/bb565b8c-48ea-11e5-88f7-c2346a645018.png">

**MacOS X**:

Note that MacOS X does take advantage of the HiDPI display, though the _icons_ are still low-res

<img width="400" alt="osx" src="https://cloud.githubusercontent.com/assets/314716/9424359/12ac4b9a-48ea-11e5-942b-16f64f5df12b.png">

Interestingly, the low resolution pixels in the PyQt5 example look crisper which suggests that it might be rendered at HiDPI but the content of the canvas is low resolution.

So just to clarify, I think there are two issues here:
- The icons themselves are not HiDPI-friendly (as seen in the MacOS X screenshot) [update: have opened an issue for that: #4978]
- Matplotlib is not taking advantage of the HiDPI with the PyQt5 canvas

**Comment by WeatherGod** [comment]
Is this in master or the latest release? I know some work has been done at
least in the macosx backend, and I could have sworn we fixed a few retina
issues in other backends. Not to say that we support retina everywhere, but
rather that we fixed some rendering issues on retina displays (things were
showing up a 4x the normal size and such).

As for the icons, it might be that they aren't high resolution enough to
really fare much better?

**Comment by astrofrog** [comment]
@WeatherGod - this is in master. The MacOS X backend does look nicer, though the *icons* there are still low DPI. I've updated the issue above to make this clearer and made the screenshots larger to show part of the canvas.

**Comment by tacaswell** [comment]
As I understand it, part of the difficulty here is that there seem to be several layers trying to compensate for this issue.  

On my laptop (linux and qt4agg) which has a stupidly high resolution screen the default settings render so that text is almost unreadable, but setting `rcparams['figure.dpi'] = 150` seems to work well.  Why this isn't the actual dpi of the screen or if this will work on mac + qt5 I don't know.

I am pretty sure that under the hood of the *Agg backends we are copying from the Agg pixel buffer to what ever the backends version of a screen buffer is (so we have per-pixel control and the gui doesn't do any interpolation).  We really should sort out how to query the gui frameworks for things like screen DPI and update the `figure.dpi` parameter accordingly.

**Comment by astrofrog** [comment]
Just for the record, if I set ``rcparams['figure.dpi'] = 150`` on Mac with Qt5, it just makes the figure larger rather than actually improving the resolution.

**Comment by mdehoon** [comment]
It's easy to fix this for the MacOSX backend, provided that the image is available at a high enough resolution (either as a .png or as a .pdf).

**Comment by mjsu** [comment]
It would be great to fix this. I wish I had the skills to do so.

**Comment by astrofrog** [comment]
@tacaswell - so just to make sure I understand, on Linux with PyQt5, if you do:

```
rcParams['figure.dpi'] = 150
plt.figure()
plt.subplot(1,1,1)
```

you end up with a figure at full DPI resolution with smooth text, and the figure is no larger than it is with the lower DPI?

I looked at the backend code for Qt5 but couldn't see where there might be a Mac/Linux difference

**Comment by tacaswell** [comment]
No, on a high dpi screen my figures show up as tiny, increasing the DPI makes the font readable.  I think that this is a result of my system not trying to compensate for the high dpi by bumping up the mpl dpi, it comes closer to what it really is, so a 5x6 figure really is 5x6 on the screen.

**Comment by astrofrog** [comment]
@tacaswell - ok, so it seems the machinery *should* be in place to get high resolution canvases, but it is just buggy on Mac, right?

Can anyone confirm whether they see the same behavior as me on Mac?

**Comment by astrofrog** [comment]
I think I found a path to a fix. I can work on it during the week but if anyone else knows the exact fix, please let me know!

**Comment by mjsu** [comment]
@astrofrog I get the same behaviour as you using the Qt5Agg backend and that rcParams. If I use the MacOSX backend with dpi=150 the figure is also bigger (and resolution is good), but the aspect ratio of the enlarged figure is different to the Qt5Agg.

**Comment by astrofrog** [comment]
I've been making progress on this, but in the process got very confused about the DPI issue and the fact that on most computers I've tried, DPI doesn't mean actual DPI but is just a multiplicative factor for figure size. So basically the question is whether a figure with a DPI of 80 should look crisp by default with PyQt5. I've sent a message for discussion to the dev list: https://mail.python.org/pipermail/matplotlib-devel/2015-September/000097.html

**Comment by johan12345** [comment]
On Windows, using the TkAgg backend, Both the plotting canvas and the icons in the toolbar have low resolution, too :(
I am using the current 1.5.1 release, Python 3.5.1 and Windows 10.

**Comment by piannucci** [comment]
Here's a patch.  Doesn't fix the icons, just the plots.  #6389

**Comment by QuLogic** [comment]
Fixed by #6387 / #7507.
